ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/api/WriteServiceWithDeduplicationOffsetSupportSpec.scala

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.api
+
+import java.time.Duration
+
+import akka.stream.Materializer
+import com.daml.error.ContextualizedErrorLogger
+import com.daml.ledger.TestLoggers
+import com.daml.ledger.api.DeduplicationPeriod
+import com.daml.ledger.api.domain.ApplicationId
+import com.daml.ledger.api.testing.utils.AkkaBeforeAndAfterAll
+import com.daml.ledger.configuration.Configuration
+import com.daml.ledger.participant.state.kvutils.deduplication.DeduplicationPeriodService
+import com.daml.ledger.participant.state.v2.SubmissionResult.Acknowledged
+import com.daml.ledger.participant.state.v2.{
+  SubmissionResult,
+  SubmitterInfo,
+  TransactionMeta,
+  WriteService,
+}
+import com.daml.lf.crypto.Hash
+import com.daml.lf.data.Time.Timestamp
+import com.daml.lf.data.{ImmArray, Ref}
+import com.daml.lf.transaction.{SubmittedTransaction, TransactionVersion, VersionedTransaction}
+import com.daml.logging.LoggingContext
+import com.daml.telemetry.{NoOpTelemetryContext, TelemetryContext}
+import org.mockito.{ArgumentMatchersSugar, MockitoSugar}
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AsyncWordSpec
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.jdk.FutureConverters._
+
+class WriteServiceWithDeduplicationOffsetSupportSpec
+    extends AsyncWordSpec
+    with MockitoSugar
+    with Matchers
+    with AkkaBeforeAndAfterAll
+    with TestLoggers
+    with ArgumentMatchersSugar {
+  implicit val telemetryContext: NoOpTelemetryContext.type = NoOpTelemetryContext
+  private val mockWriteService: WriteService = mock[WriteService]
+  private val mockDeduplicationPeriodService: DeduplicationPeriodService =
+    mock[DeduplicationPeriodService]
+  val service = new WriteServiceWithDeduplicationOffsetSupport(
+    mockWriteService,
+    mockDeduplicationPeriodService,
+  )
```

Comment:
Mocks are mutable. Please move them into the test case so we don't accidentally use them twice.

Reply:
Actually, I would say that the combo of (one test + just checking that it delegates to a mock) tells us that this class is superfluous; we can merge it with `DeduplicationPeriodService` with no loss of functionality or testability.

Reply:
Added a reset for the mocks just in case. I'm not a big fan of having them in the test case as it adds too much boilerplate to the test case itself.

I would still prefer to keep these 2 separate. If you do disagree with it I'm all up for merging them, however, my preference is because of the following reasons:
- the WriteServiceWithDeduplicationOffsetSupport (I hate the naming) is a light decorator which has a simple, restricted scope, overwrite the deduplication period with whatever it gets from the deduplication period service (as it's reflected in the test as well). This way the logic as to which deduplication period is supported is not part of the write service.  I would suggest removing the offset from the name so that it reflects this.
- in the DeduplicationPeriodService we also isolate the validation, therefore there is no validation leak into the write service.

I'm not a big fan of the mock usage as well, but it just seemed more straightforward in this case

Reply:
I'm OK with sticking to your preference, with the above suggested changes.